lock.rst

```diff
@@ -0,0 +1,291 @@
+
+                    <framework:resource>pgsql:host=127.0.0.1;dbname=lock</framework:resource>
+
+                    <framework:resource>sqlsrv:server=localhost;Database=test</framework:resource>
+
+                    <framework:resource>oci:host=localhost;dbname=test</framework:resource>
+
+                    <framework:resource>%env(LOCK_DSN)%</framework:resource>
+
+                    <!-- named locks -->
+                    <framework:resource name="invoice">semaphore</framework:resource>
+                    <framework:resource name="invoice">redis://r2.docker</framework:resource>
+                    <framework:resource name="report">semaphore</framework:resource>
+                </framework:lock>
+            </framework:config>
+        </container>
+
+    .. code-block:: php
+
+        // config/packages/lock.php
+        $container->loadFromExtension('framework', [
+            'lock' => null,
+            'lock' => 'flock',
+            'lock' => 'flock:///path/to/file',
+            'lock' => 'semaphore',
+            'lock' => 'memcached://m1.docker',
+            'lock' => ['memcached://m1.docker', 'memcached://m2.docker'],
+            'lock' => 'redis://r1.docker',
+            'lock' => ['redis://r1.docker', 'redis://r2.docker'],
+            'lock' => 'zookeeper://z1.docker',
+            'lock' => 'zookeeper://z1.docker,z2.docker',
+            'lock' => 'sqlite:///%kernel.project_dir%/var/lock.db',
+            'lock' => 'mysql:host=127.0.0.1;dbname=lock',
+            'lock' => 'pgsql:host=127.0.0.1;dbname=lock',
+            'lock' => 'sqlsrv:server=localhost;Database=test',
+            'lock' => 'oci:host=localhost;dbname=test',
+            'lock' => '%env(LOCK_DSN)%',
+
+            // named locks
+            'lock' => [
+                'invoice' => ['semaphore', 'redis://r2.docker'],
+                'report' => 'semaphore',
+            ],
+        ]);
+
+Locking a Resource
+------------------
+
+To lock the default resource, autowire the lock using
+:class:`Symfony\\Component\\Lock\\LockInterface` (service id ``lock``)::
+
+    // src/Controller/PdfController.php
+    namespace App\Controller;
+
+    use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+    use Symfony\Component\Lock\LockInterface;
+
+    class PdfController extends AbstractController
+    {
+        /**
+         * @Route("/download/terms-of-use.pdf")
+         */
+        public function downloadPdf(LockInterface $lock, MyPdfGeneratorService $pdf)
+        {
+            $lock->acquire(true);
+
+            // heavy computation
+            $myPdf = $pdf->getOrCreatePdf();
+
+            $lock->release();
+
+            // ...
+        }
+    }
+
+.. caution::
+
+    The same instance of ``LockInterface`` won't block when calling `acquire``
+    multiple times. If several services share the same ``lock`` service, they
+    won't lock each other, use ``LockFactory`` instead.
```

Comment:
I don't really understand this caution. The above example, when run concurrently *will* block eachother, right?

Reply:
Depends, what you call `concurrently`, when running currently in the same process (think react-php), the same instance of Lock (think shared service) wont block.

Bellow case where lock does not apply
```php
$lock = new Lock();
$lock->acquire(); 
$lock->acquire(); // no lock (same instance)
```

```php
$lock = new Lock();
$a = new ReactControler($lock); 
$b = new ReactControler($lock); // Both $a and  share same lock instance and wont block each other
```

Reply:
I would appreciate help to rephrase my sentence :)

Reply:
reworded

Reply:
I've revisited this caution in 1170a7c59 . Can you please check if the new contents of this caution are correct?

Reply:
sounds good.
Thanks for your help